zephyr/lib/regions_mm.c

```diff
@@ -1,52 +1,759 @@
+			 * 0 for individually allocated block and 0 for block ending allocation
+			 * Continuing the example above
+			 * First: 1111100000000000 allocate another block 1111100000000000
+			 * Second:1110000000000000 after the first one	  1110000000000000
+			 * based on information above we know exactly how long allocations were
+			 * by checking second bit array against first one
+			 * First: 1111111100000000 allocate another block 1111111100000000
+			 * Second:1110000000000000 and another two blocks 1110001000000000
+			 * We are still able to pinpoint allocations size
+			 */
+
+			/* Set bits count - 1 on the same offset as in allocation bit array */
+			allocation_bitarray_offset = (uintptr_t)ptr
+					- (uintptr_t)heap->physical_blocks_allocators
+						[mem_block_iterator]->buffer;
+			allocation_bitarray_position = allocation_bitarray_offset / block_size;
+
+			/* Need to set bits only if we have more than 1 block to allocate */
+			if (block_count - 1)
+				sys_bitarray_set_region(heap->allocation_sizes[mem_block_iterator],
+					block_count - 1, allocation_bitarray_position);
 			break;
-		case MEM_REG_ATTR_OPPORTUNISTIC_MEMORY:
-			vm_heaps[i].memory_caps = SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_HP;
+		}
+	}
+
+	/* If ptr is NULL it means we did not allocate anything and we should
+	 * break execution here
+	 */
+	if (!ptr)
+		return NULL;
+
+	/* Now we need to check if have mapped physical page already
+	 * We can do it by working out if there was any other allocation
+	 * done in the allocator section that is in CONFIG_MM_DRV_PAGE_SIZE chunk.
+	 * Start by getting start of allocation
+	 * ptr aligned to physical pages boundary
+	 */
+	phys_aligned_ptr = ALIGN_DOWN((uintptr_t)ptr, CONFIG_MM_DRV_PAGE_SIZE);
+	phys_aligned_alloc_end = ALIGN_UP((uintptr_t)ptr
+		+ alloc_size, CONFIG_MM_DRV_PAGE_SIZE);
+
+	/* To check if there was any allocation done before on block we need
+	 * to perform check operations however
+	 * we just made allocation there. We will clear it
+	 * from bit array of allocator for ease of operation.
+	 * RFC if someone knows a better way to handle it in bit array API.
+	 */
+
+	/* Clear bits of the allocation that was done */
+	sys_bitarray_clear_region(heap->physical_blocks_allocators[mem_block_iterator]->bitmap,
+		block_count, allocation_bitarray_position);
+
+	/* Now that we cleared the allocation we just made we can check
+	 * if there was any other allocation in the space that would need to be
+	 * mapped. Since this API is only one that uses mapping in specified address
+	 * ranges we can assume that we made mapping for the space if it has any previous
+	 * allocations. We check that iterating over all physical allocation blocks
+	 */
+	physical_block_count = (phys_aligned_alloc_end
+		- phys_aligned_ptr) / CONFIG_MM_DRV_PAGE_SIZE;
+
+	for (i = 0; i < physical_block_count; i++) {
+		/* Check if beginning of physical block is misaligned with
+		 * allocation block size (ie block spans two physical pages)
```

Comment:
For smaller than page blocks this is impossible as long as your block sizes are powers of 2. I think we should use this simplification: block sizes should be powers of 2 and aligned on a block-size border. Then only multi-page blocks can span multiple pages.

Reply:
resolving since this is the same issue stated above that we can skipp  the calculations for border allocations - will return to that in that comment

Reply:
but an allocation block smaller than a page size cannot cross a page border, and larger than a page, will always contain multiple whole pages, right? So at least your comment should be updated to clarify that